packages/react-dom/src/client/ReactDOMInput.js

```diff
@@ -212,35 +240,66 @@ export function postMountWrapper(
 ) {
   const node = ((element: any): InputWithWrapperState);
 
+  // Do not assign value if it is already set. This prevents user text input
+  // from being lost during SSR hydration.
   if (props.hasOwnProperty('value') || props.hasOwnProperty('defaultValue')) {
+    const type = props.type;
+    const isButton = type === 'submit' || type === 'reset';
+
     // Avoid setting value attribute on submit/reset inputs as it overrides the
     // default value provided by the browser. See: #12872
-    const type = props.type;
-    if (
-      (type === 'submit' || type === 'reset') &&
-      (props.value === undefined || props.value === null)
-    ) {
+    if (isButton && (props.value === undefined || props.value === null)) {
       return;
     }
 
-    const initialValue = toString(node._wrapperState.initialValue);
-    const currentValue = node.value;
+    const initialValue = node._wrapperState.initialValue;
 
     // Do not assign value if it is already set. This prevents user text input
     // from being lost during SSR hydration.
     if (!isHydrating) {
-      // Do not re-assign the value property if there is no change. This
-      // potentially avoids a DOM write and prevents Firefox (~60.0.1) from
-      // prematurely marking required inputs as invalid
-      if (initialValue !== currentValue) {
-        node.value = initialValue;
+      if (disableInputAttributeSyncing) {
+        // When not syncing the value attribute, the value property points
+        // directly to the React prop. Only assign it if it exists.
+        if (props.hasOwnProperty('value')) {
+          const value = getToStringValue(props.value);
+
+          // Always assign on buttons so that it is possible to assign an
+          // empty string to clear button text.
+          //
+          // Otherwise, do not re-assign the value property if is empty. This
+          // potentially avoids a DOM write and prevents Firefox (~60.0.1) from
+          // prematurely marking required inputs as invalid. Equality is compared
+          // to the current value in case the browser provided value is not an
+          // empty string.
+          if (isButton || value !== node.value) {
+            node.value = toString(value);
+          }
+        }
+      } else {
+        // When syncing the value attribute, the value property should use
+        // the the wrapperState._initialValue property. This uses:
+        //
+        //   1. The value React property when present
+        //   2. The defaultValue React property when present
+        //   3. An empty string
+        if (initialValue !== node.value) {
```

Comment:
This used to compare `toString(node._wrapperState.initialValue) !== node.value`.

But now it compares `node._wrapperState.initialValue !== node.value`.

Is that change safe?

Reply:
You are correct. This should call toString()

Reply:
Needs a regression test?

Reply:
Hard. We only need to call toString for value comparison so that we do not reassign the existing default value and trigger validation. `null` and `undefined` are already skipped. This also happens on the post mount step, and not during hydration. I'm not sure 

We have a manual test fixture for this reason (sorry, I know the formatting isn't ideal):
https://github.com/facebook/react/blob/a7bd7c3c0480c96e32d9518bb707172dfc72db58/fixtures/dom/src/components/fixtures/text-inputs/index.js#L65-L99